packages/protocol/test/governance/election.ts

```diff
@@ -1212,4 +1212,319 @@ contract('Election', (accounts: string[]) => {
+          const index = 0
+          const slashedValue = value
+          const remaining = value - slashedValue
+          beforeEach(async () => {
+            await election.slashVotes(voter, slashedValue, [NULL_ADDRESS], [NULL_ADDRESS], [index])
+          })
+
+          it('should decrement pending votes to zero', async () => {
+            assertEqualBN(await election.getPendingVotesForGroupByAccount(group, voter), remaining)
+          })
+
+          it('should decrement total votes to zero', async () => {
+            assertEqualBN(await election.getTotalVotesForGroupByAccount(group, voter), remaining)
+            assertEqualBN(await election.getTotalVotesByAccount(voter), remaining)
+            assertEqualBN(await election.getTotalVotesForGroup(group), remaining)
+            assertEqualBN(await election.getTotalVotes(), remaining)
+          })
+
+          it("should increment `voter`'s nonvoting account balance", async () => {
+            assertEqualBN(await mockLockedGold.nonvotingAccountBalance(voter), slashedValue)
+          })
+
+          it("should remove the group from the voter's voted set", async () => {
+            assert.deepEqual(await election.getGroupsVotedForByAccount(voter), [])
+          })
+        })
+      })
+
+      describe('when the account only has active votes', () => {
+        beforeEach(async () => {
+          await mineBlocks(EPOCH, web3)
+          await election.activate(group)
+        })
+
+        describe('when the account is slashed for the total active voting gold', () => {
+          const index = 0
+          const slashedValue = value
+          const remaining = value - slashedValue
+          beforeEach(async () => {
+            await election.slashVotes(voter, slashedValue, [NULL_ADDRESS], [NULL_ADDRESS], [index])
+          })
+
+          it('should decrement active voted gold to zero', async () => {
+            assertEqualBN(await election.getActiveVotesForGroupByAccount(group, voter), remaining)
+          })
+
+          it('should decrement total voted gold to zero', async () => {
+            assertEqualBN(await election.getTotalVotesForGroupByAccount(group, voter), remaining)
+            assertEqualBN(await election.getTotalVotesByAccount(voter), remaining)
+            assertEqualBN(await election.getTotalVotesForGroup(group), remaining)
+            assertEqualBN(await election.getTotalVotes(), remaining)
+          })
+
+          it("should increment `voter`'s nonvoting account balance", async () => {
+            assertEqualBN(await mockLockedGold.nonvotingAccountBalance(voter), slashedValue)
+          })
+
+          it("should remove the group from the voter's voted set", async () => {
+            assert.deepEqual(await election.getGroupsVotedForByAccount(voter), [])
+          })
+        })
+      })
+    })
+
+    describe('when the account has voted for more than one group equally', () => {
```

Comment:
There seems to be a decent amount of duplicated code in the following 3 test cases. Would it be possible to define something that parses objects like the following, sets up the test case, and then asserts the expected values?


```
const testCases = [
  {
    votesBefore: [
       {
          group: 0x1          
          pending: 2000
          active: 500
       }, {
          group: 0x2         
          pending: 1000
          active: 0
       }
    ],
    slash: 2700
    votesAfter: [
       {
          group: 0x1          
          pending: 0
          active: 0
       }, {
          group: 0x2         
          pending: 800
          active: 0
       }
    ]
  }]
}  
```

Reply:
Up to you if you want to delay merging for this, I'll get started on it once I finish addressing reviews for #2317 since it seems lower priority, but still worth adding.

Reply:
Not worth blocking the PR but something I suggest we do after since it will clean up the testing code quite a bit